### PR TITLE
Add versioning capability to pynautobot

### PR DESCRIPTION
**Pull request by timizuoebideri1**
- Add a global api_version argument on initalization of API class
 - Add versioning per specific request like get(), list(), create() e.t.c

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -96,6 +96,31 @@ nautobot = pynautobot.api(
...
+
+Used for Nautobot Rest API versioning. Versioning can be controlled globally by setting `api_version` on initialization of the `API` class and/or for a specific request e.g (`list()`, `get()`, `create()` etc.) by setting an optional `api_version` parameter.
+
+__Global versioning__
+```python
+import pynautobot
+nautobot = pynautobot.api(
+    url="http://localhost:8000",
+    token="d6f4e314a5b5fefd164995169f28ae32d987704f",
+    api_version="1.3
```
Comment (glennmatthews) [inline README.md:109]:
```suggestion
    api_version="1.3"
```

File: pynautobot/core/api.py [hunk 2]
```diff
@@ -54,6 +54,8 @@ class Api(object):
     :param str token: Your Nautobot token.
     :param bool,optional threading: Set to True to use threading in ``.all()``
         and ``.filter()`` requests.
+    :param str, optional api_version: Set to overide the default Nautobot Rest API Version
+        for all request.
```
Comment (glennmatthews) [inline pynautobot/core/api.py:58]:
```suggestion
    :param str,optional api_version: Set to override the default Nautobot REST API Version
        for all requests.
```

File: pynautobot/core/endpoint.py [hunk 3]
```diff
@@ -76,10 +76,13 @@ def _lookup_ret_obj(self, name, model):
...
         return ret
 
-    def all(self):
+    def all(self, api_version=None):
         """Queries the 'ListView' of a given endpoint.
 
         Returns all objects from an endpoint.
+        
+        :arg str, optional api_version: Overide default or globally set Nautobot Rest API 
+            version for this(all) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:85]:
```suggestion
        :arg str,optional api_version: Override default or globally-set Nautobot REST API 
            version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 4]
```diff
@@ -107,6 +113,9 @@ def get(self, *args, **kwargs):
         :arg str,optional \**kwargs: Accepts the same keyword args as
             filter(). Any search argument the endpoint accepts can
             be added as a keyword arg.
+        
+        :arg str, optional api_version: Overide default or globally set Nautobot Rest API 
+            version for this(get) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:118]:
```suggestion
        :arg str,optional api_version: Override default or globally-set Nautobot REST API 
            version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 5]
```diff
@@ -168,6 +181,8 @@ def filter(self, *args, **kwargs):
             accepted on given endpoint.
         :arg str,optional \**kwargs: Any search argument the
             endpoint accepts can be added as a keyword arg.
+        :arg str, optional api_version: Overide default or globally set 
+            Nautobot Rest API version for this(filter) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:185]:
```suggestion
        :arg str,optional api_version: Override default or globally-set 
            Nautobot REST API version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 6]
```diff
@@ -233,6 +250,8 @@ def create(self, *args, **kwargs):
             properties of the objects to be created.
         :arg str \**kwargs: key/value strings representing
             properties on a json object.
+        :arg str, optional api_version: Overide default or globally set 
+            Nautobot Rest API version for this(create) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:254]:
```suggestion
        :arg str,optional api_version: Override default or globally-set 
            Nautobot REST API version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 7]
```diff
@@ -268,20 +287,25 @@ def create(self, *args, **kwargs):
...
         """Returns all choices from the endpoint.
 
         The returned dict is also saved in the endpoint object (in
         ``_choices`` attribute) so that later calls will return the same data
         without recurring requests to Nautobot. When using ``.choices()`` in
         long-running applications, consider restarting them whenever Nautobot is
         upgraded, to prevent using stale choices data.
+        
+        :arg str, optional api_version: Overide default or globally set 
+            Nautobot Rest API version for this(choices) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:308]:
```suggestion
        :arg str,optional api_version: Override default or globally-set 
            Nautobot REST API version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 8]
```diff
@@ -331,6 +359,8 @@ def count(self, *args, **kwargs):
             accepted on given endpoint.
         :arg str,optional \**kwargs: Any search argument the
             endpoint accepts can be added as a keyword arg.
+        :arg str, optional api_version: Overide default or globally set 
+            Nautobot Rest API version for this(count) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:363]:
```suggestion
        :arg str,optional api_version: Override default or globally-set 
            Nautobot REST API version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 9]
```diff
@@ -371,27 +405,36 @@ def __init__(self, parent_obj, name, custom_return=None):
...
+            http_session=parent_obj.api.http_session,
+        )
+
+    def list(self, api_version=None, **kwargs):
         r"""The view operation for a detail endpoint
 
         Returns the response from Nautobot for a detail endpoint.
 
         Args:
+            api_version str, optional: Overide default or globally set Nautobot Rest API version for this(list) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:421]:
```suggestion
            :arg str,optional api_version: Override default or globally set Nautobot REST API version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 10]
```diff
@@ -400,11 +443,16 @@ def create(self, data=None):
             key/value pair of the items you're creating on the parent
             object. Defaults to empty dict which will create a single
             item with default values.
+        :args api_version str, optional: Overide default or globally set 
+            Nautobot Rest API version for this(create) request.
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:447]:
```suggestion
        :args str,optional api_version: Override default or globally set 
            Nautobot REST API version for this single request.
```

File: pynautobot/core/endpoint.py [hunk 11]
```diff
@@ -371,27 +405,36 @@ def __init__(self, parent_obj, name, custom_return=None):
...
             **kwargs: key/value pairs that get converted into url parameters when passed to the endpoint.
                 E.g. ``.list(method='get_facts')`` would be converted to ``.../?method=get_facts``.
 
         :returns: A dictionary or list of dictionaries retrieved from
             Nautobot.
         """
+        if api_version:
+            self.request_kwargs["api_version"] = api_version
+        
         req = Request(**self.request_kwargs).get(add_params=kwargs)
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:431]: Need to default to parent_obj.api.api_version, I think - also modifying self.request_kwargs seems wrong as that will persist to future calls on this same endpoint.

```suggestion
        api_version = api_version or self.parent_obj.api.api_version
        
        req = Request(api_version=api_version, **self.request_kwargs).get(add_params=kwargs)
```

File: pynautobot/core/endpoint.py [hunk 12]
```diff
@@ -400,11 +443,16 @@ def create(self, data=None):
...
+            Nautobot Rest API version for this(create) request.
 
         :returns: A dictionary or list of dictionaries its created in
             Nautobot.
         """
         data = data or {}
+        if api_version:
+            self.request_kwargs["api_version"] = api_version
+            
         req = Request(**self.request_kwargs).post(data)
```
Comment (glennmatthews) [inline pynautobot/core/endpoint.py:456]:
```suggestion
        api_version = api_version or self.parent_obj.api.api_version
            
        req = Request(api_version=api_version, **self.request_kwargs).post(data)
```

File: pynautobot/core/query.py [hunk 13]
```diff
@@ -131,6 +131,7 @@ def __init__(
                 In (e.g. /api/dcim/devices/?name='test') 'name': 'test'
                 would be in the filters dict.
             key (int, optional): database id of the item being queried.
+            api_version (str, optional): Set to overide the default Nautobot Rest API Version.
```
Comment (glennmatthews) [inline pynautobot/core/query.py:134]:
```suggestion
            api_version (str, optional): Set to override the default Nautobot REST API Version.
```

**Comment by jvanderaa** [comment]
Sounds like a reasonable action, hopefully the device-type library will work. Thoughts on maybe making a fork of it?

**Comment by jvanderaa** [comment]
@timizuoebideri1 will you go back a few commits on the checkout?

**Comment by glennmatthews** [comment]
Can we add some tests for this feature? I'm thinking at least:

- Unit: Set api_version at the api level, verify that all mock requests include the correct header
- Unit: Set api_version at per-request level, verify that mock requests include the overridden value in their headers
- Integration: Set invalid api_version (e.g. `0.0`) and verify that the error response from Nautobot is handled gracefully

**Comment by jvanderaa** [comment]
@jmcgill298 tag

**Review by glennmatthews — APPROVED** [review]
Looks good to me

**Review by jmcgill298 — APPROVED** [review]

**Pull request merged**